File: go/internal/model/evidence.go
package model

import "strings"

// mechanical is the set of genuine tool-receipt source heads. A source whose
// head (before the first ":") is in this set — and which is not llm-namespaced —
// is a mechanical receipt. Ported verbatim from evidence.py:14 (the eight
// sources); this is a security control, not a feature, so the set is frozen.
var mechanical = map[string]bool{
	"semgrep":          true,
	"codeql":           true,
	"ast-grep":         true,
	"tree-sitter":      true,
	"ripgrep":          true,
	"structural-index": true,
	"secrets":          true,
	"sca":              true,
}

// Confidence is a finding's tier, derived from its strongest evidence source.
// Values match the Python Confidence enum exactly.
type Confidence string

// Confidence tiers, matching the Python Confidence enum values.
const (
	ConfidenceHigh   Confidence = "high"
	ConfidenceMedium Confidence = "medium"
	ConfidenceLow    Confidence = "low"
)

// IsToolReceipt reports whether source is a genuine mechanical-tool receipt.
//
// Any source with an "llm" prefix is rejected before the mechanical-set check,
// so an LLM assertion can never be counted as a receipt. Otherwise the head
// (before the first ":") must be one of the eight mechanical sources. Ported
// verbatim from evidence.py is_tool_receipt.
func IsToolReceipt(source string) bool {
	if strings.HasPrefix(source, "llm") {
		return false
	}
	head, _, _ := strings.Cut(source, ":")
	return mechanical[head]
}

// AsLLMClaim namespaces an LLM-asserted source so it cannot masquerade as a
// receipt: the source is returned unchanged if already "llm"-prefixed, else it
// is prefixed with "llm-claimed:". Ported verbatim from evidence.py as_llm_claim.
func AsLLMClaim(source string) string {
	if strings.HasPrefix(source, "llm") {
		return source
	}
	return "llm-claimed:" + source
}

// ConfidenceFor grades a finding from its evidence sources (strongest link):
// any real tool receipt yields High; else any "llm-corroborated"-prefixed source
// yields Medium; else Low. This is the confirm gate — an llm-claimed-only finding
// can never reach High. Ported verbatim from evidence.py confidence_for.
func ConfidenceFor(sources []string) Confidence {
	for _, s := range sources {
		if IsToolReceipt(s) {
			return ConfidenceHigh
		}
	}
	for _, s := range sources {
		if strings.HasPrefix(s, "llm-corroborated") {
			return ConfidenceMedium
		}
	}
	return ConfidenceLow
}

File: skills/sec-harness/helpers/sec_harness/evidence.py
"""Evidence grading: mechanical tool receipts outrank LLM assertions.

A finding's confidence is the strongest evidence in its chain: a real tool
receipt (semgrep/codeql/ast-grep/tree-sitter/ripgrep) yields HIGH; an LLM
assertion corroborated by nothing yields LOW. LLM-asserted evidence is
namespaced ``llm-claimed:`` so it can never be counted as a tool receipt.
Adapted from raptor's evidence_grade.
"""

from __future__ import annotations

from enum import Enum

_MECHANICAL = {"semgrep", "codeql", "ast-grep", "tree-sitter", "ripgrep",
               "structural-index", "secrets", "sca"}


class Confidence(str, Enum):
    """Confidence tier for a finding, from its strongest evidence."""

    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


def is_tool_receipt(source: str) -> bool:
    """Return True if ``source`` is a genuine mechanical-tool receipt.

    Args:
        source: An evidence source string (e.g. ``codeql:dataflow``).

    Returns:
        True only for mechanical-tool sources not namespaced ``llm-claimed:``.
    """
    if source.startswith(("llm-claimed:", "llm")):
        return False
    return source.split(":", 1)[0] in _MECHANICAL


def as_llm_claim(source: str) -> str:
    """Namespace an LLM-asserted source so it cannot masquerade as a receipt.

    Args:
        source: The raw source the LLM claims.

    Returns:
        The source unchanged if already ``llm``-prefixed, else ``llm-claimed:<source>``.
    """
    return source if source.startswith("llm") else f"llm-claimed:{source}"


def confidence_for(sources: list[str]) -> Confidence:
    """Grade a finding's confidence from its evidence sources (strongest link).

    Args:
        sources: Evidence source strings.

    Returns:
        HIGH if any real tool receipt; else MEDIUM if any ``llm-corroborated``;
        else LOW.
    """
    if any(is_tool_receipt(s) for s in sources):
        return Confidence.HIGH
    if any(s == "llm-corroborated" or s.startswith("llm-corroborated") for s in sources):
        return Confidence.MEDIUM
    return Confidence.LOW
